Handle product ecotax in product v2 (for both standard and combination form)

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/8/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR introduces the management of ecotax in the product page It automatically recomputes the tax excluded values when the ecotax is modified in order not to change the final price with tax included This behaviour is implemented in the tab pricing for all types of products And also in the product with combinations where the ecotax can be overridden In which case the impact is updated to avoid changing the final price as well
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #25907
| Related PRs       | ~
| How to test?      | Enable ecotax in the Localization > Taxes section then you can edit the ecotax in the product page for the standard value and the combination's one
| Possible impacts? | ~


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

#### BC break

- [GetCombinationForEditingHandler](https://github.com/PrestaShop/PrestaShop/pull/28482/files#diff-2a9b96c35135bfcf66b5306e8654da1b762dccce35476897fb7a5d59fc39a4c7): constructor uses Configuration service instead of $countryId
- [CombinationPrices](https://github.com/PrestaShop/PrestaShop/pull/28482/files#diff-339c07628a32c56eed150c5826cacf198e4379e311b5d14c1f3765eef08106ef): adds ne parameters and change parameter orders
- [src/Core/Domain/Product/QueryResult/ProductPricesInformation.php](https://github.com/PrestaShop/PrestaShop/pull/28482/files#diff-c7d5bec603e2ebd8fe74034ce32967edaaef16a1d0f7537e6f64b0398104de88): add $ecotaxTaxIncluded parameter
- [src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationPriceImpactType.php](https://github.com/PrestaShop/PrestaShop/pull/28482/files#diff-30c4d3f84514e964f847ad50e7c88787003726da2c3d8b7756f9aec9f095a66e): add 5 parameters to the constructor
- [src/PrestaShopBundle/Form/Admin/Sell/Product/Pricing/RetailPriceType.php](https://github.com/PrestaShop/PrestaShop/pull/28482/files#diff-b565fd5b8fb8863742ddb24f4f19a9a74a794a201dd8ed5f377d7cc616320db2): inject choice provider service instead of choices, add three other parameters to constructor

jolelievre committed:
Integrate ecotax field correctly in the form

jolelievre committed:
Handle ecotax with tax included and impact in rpduct form query

jolelievre committed:
Add behat tests for product eco tax

jolelievre committed:
Handle ecotax new structure in data provider and form handler

jolelievre committed:
Handle JS automatic update for eco tax values

jolelievre committed:
Handle more fields in combination price edition and udpated behat tests

jolelievre committed:
Integrate ecotax fields in the combination form

jolelievre committed:
Adapt combination data provider and builder to the new form structure

jolelievre committed:
Handle tax update in the case of ecotax enabled

jolelievre committed:
Add some product values in the combination form query as they will be needed for computing in the interface

jolelievre committed:
Integrate product data in hidden inputs

jolelievre committed:
Handle automatic computing in combination form of values with taxes and final price preview

jolelievre committed:
Combination use CLDR to display final price

jolelievre committed:
Use float value of tax rating everywhere instead of percent values

jolelievre committed:
Handle ecotax impact in combination form

jolelievre committed:
Corectly include the ecotax in the final displayed price

matks commented on admin-dev/themes/new-theme/js/pages/product/combination/form/combination-form-model.ts:
```diff
@@ -67,4 +84,169 @@ export default class CombinationFormModel {
+      }
+      case 'price.ecotaxTaxExcluded': {
+        // We first update the impact price before the final price is updated because it is important in the computing
+        // since updateFinalPrices is called after eahc input update it would be triggered too soon if we start by updating
```
```suggestion
        // since updateFinalPrices is called after each input update it would be triggered too soon if we start by updating
```

matks commented on src/Adapter/Form/ChoiceProvider/TaxRuleGroupChoiceProvider.php:
```diff
@@ -61,7 +61,7 @@ public function getChoicesAttributes(): array
 
             $attrs[$rule['name']] = [
-                'data-tax-rate' => !empty($rule['rate']) ? $rule['rate'] : null,
+                'data-tax-rate' => !empty($rule['rate']) ? $rule['rate'] / 100 : 0,
```
I'm not sure to understand why this is needed

jolelievre replied:
The idea is to have some convention across the whole, all the other tax rates are provided in their "float" value (between 0 and 1) instead of the percent one (between 1 and 100) because anyway the JS had to perform this operation to compute the tax ratio.

One is no better than the other in the end, but I thought it was probably better to have a common format instead of values as float and other as percent. And even if this provider is shared by two form types there is only the product one which uses the `getChoicesAttributes` method so no impact outside of this context.

jolelievre replied:
Oh but I just noticed it was I who developed the other code that was not in percentage value It's recent code so it's still modifiable, I'm gonna update it to use the percent value everywhere

matks commented on src/Core/Domain/Product/QueryResult/ProductPricesInformation.php:
```diff
@@ -51,6 +51,11 @@ class ProductPricesInformation
+    /**
+     * @var DecimalNumber
+     */
+    private $ecotaxTaxIncluded;
```
Funny name XD

jolelievre replied:
Yeah, I admit 😅. But even if it's redundant it felt like the one with less confusion `ecotaxIncluded` would not have been much clearer

matks reviewed (changes requested):
CI is red 😢 

Apart from CI I noticed a few improveable items but nothing blocking. If CI passes I will approve this Pull Request.

jolelievre committed:
Fix linter errors

jolelievre committed:
Use tax rate value as percentage everywhere for common convention and because it is the raw value from DB

jolelievre committed:
Improve precision in TaxComputer

Progi1984 commented on tests/Integration/Behaviour/Features/Context/Domain/TaxRulesGroupFeatureContext.php:
```diff
@@ -61,6 +61,17 @@ public function assertTaxRuleGroupExists(string $name)
+     *
+     * @param string $name
+     */
+    public function identifyTaxRulesGroup(string $name, string $taxRuleGroupReference)
```
```suggestion
    public function identifyTaxRulesGroup(string $name, string $taxRuleGroupReference): void
```

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Pricing/RetailPriceType.php:
```diff
@@ -79,32 +78,53 @@ class RetailPriceType extends TranslatorAwareType
         bool $taxEnabled,
-        bool $isEcotaxEnabled
+        bool $isEcotaxEnabled,
+        int $ecoTaxGroupId,
+        TaxComputer $taxComputer,
+        int $contextCountryId
```
Please define BC Breaks

jolelievre replied:
Not on product experimental code

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationPriceImpactType.php:
```diff
@@ -52,20 +58,49 @@ class CombinationPriceImpactType extends TranslatorAwareType
         Currency $defaultCurrency,
-        string $weightUnit
+        string $weightUnit,
+        bool $isEcotaxEnabled,
+        int $ecoTaxGroupId,
+        TaxComputer $taxComputer,
+        int $contextCountryId,
+        Locale $contextLocale
```
Please define BC Break

jolelievre replied:
Not on product experimental code

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationPriceImpactType.php:
```diff
@@ -132,6 +236,29 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'required' => false,
+                'label' => $this->trans('Cost price (tax excl.)', 'Admin.Catalog.Feature'),
+                'label_tag_name' => 'h3',
+                'attr' => ['data-display-price-precision' => self::PRESTASHOP_DECIMALS],
```
```suggestion
                'attr' => ['data-display-price-precision' => static::PRESTASHOP_DECIMALS],
```

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationPriceImpactType.php:
```diff
@@ -74,48 +109,117 @@ public function __construct(
+                'label' => $this->trans('Impact on price (tax excl.)', 'Admin.Catalog.Feature'),
+                'label_help_box' => $this->trans('If the price of this combination is different from the initial retail price, enter the value of the impact (negative or positive).', 'Admin.Catalog.Help'),
+                'attr' => [
+                    'data-display-price-precision' => self::PRESTASHOP_DECIMALS,
```
```suggestion
                    'data-display-price-precision' => static::PRESTASHOP_DECIMALS,
```

Progi1984 commented on src/Core/Domain/Product/QueryResult/ProductPricesInformation.php:
```diff
@@ -108,6 +114,7 @@ public function __construct(
         DecimalNumber $price,
         DecimalNumber $priceTaxIncluded,
         DecimalNumber $ecotax,
+        DecimalNumber $ecotaxTaxIncluded,
```
Please define BC Breaks

jolelievre replied:
Not on product experimental code

Progi1984 commented on src/Core/Domain/Product/Combination/QueryResult/CombinationPrices.php:
```diff
@@ -38,55 +38,77 @@ class CombinationPrices
         DecimalNumber $impactOnPriceTaxIncluded,
         DecimalNumber $impactOnUnitPrice,
-        DecimalNumber $wholesalePrice
+        DecimalNumber $impactOnUnitPriceTaxIncluded,
+        DecimalNumber $ecotax,
+        DecimalNumber $ecotaxTaxIncluded,
+        DecimalNumber $wholesalePrice,
+        DecimalNumber $productTaxRate,
+        DecimalNumber $productPrice,
+        DecimalNumber $productEcotax
```
Please define BC Breaks

jolelievre replied:
Not on product experimental code

Progi1984 commented on src/Adapter/Tax/TaxComputer.php:
```diff
@@ -29,12 +29,19 @@
+     * The conversion between tax rate as percent to tax ratio as float value can make us lose some precision,
+     * so we increase the default precision (6) to avoid losing two digits by diving by 100 (two decimal factors).
+     */
+    private const DIVISION_PRECISION = Division::DEFAULT_PRECISION + 2;
```
```suggestion
    protected const DIVISION_PRECISION = Division::DEFAULT_PRECISION + 2;
```

Progi1984 commented on src/Adapter/Tax/TaxComputer.php:
```diff
@@ -99,6 +108,19 @@ private function getTaxRatio(TaxRulesGroupId $taxRulesGroupId, CountryId $countr
+     *
+     * @return DecimalNumber
+     */
+    private function getTaxRatio(TaxRulesGroupId $taxRulesGroupId, CountryId $countryId): DecimalNumber
```
```suggestion
    protected function getTaxRatio(TaxRulesGroupId $taxRulesGroupId, CountryId $countryId): DecimalNumber
```

Progi1984 commented on src/Adapter/Product/Combination/QueryHandler/GetCombinationForEditingHandler.php:
```diff
@@ -129,7 +131,7 @@ public function __construct(
         TaxComputer $taxComputer,
         int $contextLanguageId,
-        int $countryId
+        Configuration $configuration
```
Please define BC Break

Progi1984 reviewed (commented):
Some feedbacks (documentation, scope for module inheritance) but it's a good job

jolelievre committed:
Handle PR feedback

khouloudbelguith reviewed (commented):
Hi @jolelievre,

I have some issues:

**First issue**

1. Go to BO > International > Taxes > Enable ecotax without tax rule
2. Go to BO > Catalog > Product V2
3. Add a simple product with Retail price tax excluded = 10 / tax rule = 20% / Retail price tax included = 12
4. Add an ecotax = 2 & save 
![image](https://user-images.githubusercontent.com/16067358/168650109-0a9cbc7a-c91f-4142-acfb-3caeae482b72.png)
5. Go to Products V2 list
6. See error: the price tax included is NOK 
![image](https://user-images.githubusercontent.com/16067358/168650254-887248c3-d244-49e0-9557-7ad5b38f44cc.png)

The price tax excluded = 8.33, I think is NOK also, because it should be 10.33 as displayed in the FO?
![image](https://user-images.githubusercontent.com/16067358/168650575-34ad8dd7-ab89-4243-a1bf-52ca1848fab9.png)
7. Change the type of the product t Product with combinations
8. Generate combinations
9. See error: in the combinations list `Final price (tax excl.)` is NOK 
![image](https://user-images.githubusercontent.com/16067358/168650886-a9c4ac26-3b01-4336-b5cd-17bac17a5b71.png)

**Second issue**

1. Go to BO > International > Taxes > Enable ecotax without tax rule
2. Go to BO > Catalog > Product V2
3. Create a product with combinations > In the pricing tab: tax rule = 20% Retail price = 0 / ecotax = 0 )
4. Generate a combination & edit it > impact on price tax included = 12 / impact on price tax excluded = 10
5. Add an ecotax = 1 & save
![image](https://user-images.githubusercontent.com/16067358/168651638-f27b448f-0aa0-4aac-80e2-e8ce0bf2ddb0.png)
6. See error in the combination list > `Final price (tax excl.)` is NOK 
![image](https://user-images.githubusercontent.com/16067358/168651803-db878c03-cdc0-4f7b-b527-54be38f57f17.png)

**Third issue**
1. Create a tax rule = 23%
2. Go to BO > Catalog > Product V2
3. Create a product with combinations > In the pricing tab: tax rule = 23% Retail price tax secluded = 10 / Retail price tax included = 12.3 / ecotax = 0 )
4. Generate a combination & edit it > impact on price tax included = 12 and click outside
5. See error: it will be converted to `12.000001`

https://user-images.githubusercontent.com/16067358/168652698-82d4d900-f087-440c-9541-97555031f4c1.mp4


**Fourth issue**
2. Go to BO > Catalog > Product V2
3. Create a product with combinations
4. Generate a combination & edit it 
5. Save
6. See error : the alert  `Update successful` is not well displayed 
![image](https://user-images.githubusercontent.com/16067358/168654565-4441f8b0-7896-4266-98a8-2c7effa3870f.png)

**_WithoutPr_** => ok
![image](https://user-images.githubusercontent.com/16067358/168654664-b259dafa-62c9-4c0a-8d86-459af3d8dd75.png)


**Fifth issue**

1. Go to BO > International > Taxes & Enable Ecotax & tax ecotax = 20%
2. Create a product with ecotax, in my case   
![image](https://user-images.githubusercontent.com/16067358/168655282-215517db-c3c3-498d-8e96-ebcafa541b32.png)
3. Go to BO > Catalog > Cart rules page and create a cart rule:
3.1 Product Selection: select the ecotax product 
![image](https://user-images.githubusercontent.com/16067358/168655072-1c7456a1-f9e8-4a0b-9eda-d3e091c2f867.png)
3.2 a percent discount, in my case 50% & Apply a discount to Selected Products 
![image](https://user-images.githubusercontent.com/16067358/168655104-c956f6d1-ed8d-44b0-b0c8-075cd9c4c03c.png)
4. Go to FO > Add the last product created to the cart
5. Apply the discount
6. See error: the discount line = 5.80 => NOK, it should be 6
![image](https://user-images.githubusercontent.com/16067358/168655420-28c90d29-0979-40b1-80f5-667f8407f94e.png)


If we place the order => the discount is ok on the Order details page 
![image](https://user-images.githubusercontent.com/16067358/168655566-6265d033-e88b-4cbd-80d8-d324156ac8d7.png)

Thanks!

jolelievre committed:
Handle computing price with ecotax in it in the product list

jolelievre commented:
Hi @khouloudbelguith 

thank you for the feedback.

#### Issue 1
Indeed the ecotax was not handled in the product page, it's not the scope of this PR which focuses on the form, but since it was not too long to handle it and too much code I added the feature in this PR.

#### Issue 2
Indeed the ecotax has not been handled in the combination list yet, it is not the scope of this PR. For this one I tried to add it quickly but it's much more complicated that the product list because it's full ajax and JS and requires too many changes. I don't want to explode the scope of this PR so this one will have to be handled in a dedicated issue for the combination list. It will probably imply a refacto, because the computation code in the list will be very similar to the one in the form. So the idea is to first validate the behaviour in the form to be sure the computing is correct and then apply it in the list (in another PR) since it also involves some editable inputs (unlike product list which is why it was simpler).

#### Issue 3
This is an old bug related to float lack of precision common to every programming language, we have the exact same behaviour in the legacy page so it won't be fixed in this PR.

#### Issue 4
Actually, the former UX integration was not correct we are not supposed to have padding on this form so I removed it as a bonus in this PR. That said the current UX is not the final one either, we need to improve the whole form for combination but it's not the scope of this PR so it won't be fixed now.

#### Issue 5
This bug is not related to the product page from BO, it seems like a bug in the cart computing in FO (and maybe in BO as well since they share the same code). The same bug happens when we do the same thing in the legacy page so it won't be fixed in this PR. We should open a dedicated issue for this bug though (unless it already exists).

jolelievre committed:
Fix NaN in previews when input are empty

jolelievre commented:
Hi again @khouloudbelguith 

however, since the scope is related, I just fixed a bug that @HanaRebaiQA indicated to me when the price input fields are empty (or `.`) to avoid displaying `NaN` in the previews.

![image (2)](https://user-images.githubusercontent.com/13801017/168857022-f5cda7db-5e83-49f5-8f45-a8b6344063c2.png)
![image (3)](https://user-images.githubusercontent.com/13801017/168857105-9cb8cd0e-bab3-476a-8c19-bf2a680c119b.png)

This is now fixed.

Progi1984 commented on src/Adapter/Product/Grid/Data/Factory/ProductGridDataFactoryDecorator.php:
```diff
@@ -91,7 +106,10 @@ public function __construct(
         int $countryId,
         ProductImagePathFactory $productImagePathFactory,
-        TranslatorInterface $translator
+        TranslatorInterface $translator,
+        bool $taxEnabled,
+        bool $isEcotaxEnabled,
+        int $ecoTaxGroupId
```
Could you document the BC Break ?

Progi1984 reviewed (dismissed):
I approve. But it lacks a BC Break to document :)

MatShir commented:
I'm ok with @jolelievre comments, I will create the issues.

jolelievre committed:
Display correct final price in header when ecotax is set

jolelievre committed:
When ecotax is enabled we must use for unit price and tax ratio impacts

jolelievre committed:
Handle ecotax even when tax is disabled in PHP code

jolelievre committed:
Fix bug related to unit price reset

jolelievre committed:
Fix unit price summary

jolelievre committed:
Add proper currency symbol for ecotax fields, and in combination list

jolelievre committed:
Improve hasSameValue for numbers using BigNumber

jolelievre committed:
Safe guard all prices input to make sure they always have a valid value if not fallback to zero

jolelievre committed:
Clean removed protected field

jolelievre committed:
Reset ecotax value on virtual product on switch

jolelievre committed:
Update type listener to handle new ecotax fields names

jolelievre committed:
Add warning on switch and handle error due to missing/extra field

jolelievre committed:
When we switch to virtual type the ecotax value is added back to the product price tax excluded

jolelievre committed:
Integrate inital product type in the form, which allows more accurate rules in the form listener to remove fields Especially during a type transition request

jolelievre committed:
Fix test after rebase

khouloudbelguith reviewed (commented):
Hi @jolelievre,

**First issue**

1. enable ecotax no tax
2. create a simple product V2
3. Set retail price tax excluded = 10 / no tax rule / retail price tax included = 10 / ecotax = 1
4. Save
5. Go to BO > Catalog > Product list V2
6. in the field `Price (tax excl.)` **Min**, Set the price of the last product created = 10
8. Click on the search button
9. See an error, the product is not displayed in the list

If I set the field `Price (tax excl.)` **Min** = 9 and I search => The product is well displayed

I attached a screen record

https://user-images.githubusercontent.com/16067358/170456062-034055e6-b4fa-477b-8e15-f01439410e0c.mp4

Also, the sort is impacted, the result of the sort by `Price (tax excl.)` will be NOK
I attached a screen record

https://user-images.githubusercontent.com/16067358/170458438-1186bdec-b085-4aa8-9114-7b5239760845.mp4

**Second issue**
1. Enable ecotax
2. Enable Multistore
3. Create a new shops
4. Go to BO > Catalog > Product V2
5. Edit a product
6. In the pricing tab
7. Edit the ecotax tax **excluded** > checkbox is displayed
8. Click on it 
9. See error, the check of the Ecotax tax included should be automatically clicked

In the other case, this issue may cause another issue following the steps
1. Enable ecotax
2. Enable Multistore
3. Create a new shops
4. Go to BO > Catalog > Product V2
5. Edit a product
6. In the pricing tab
7. Edit the ecotax tax **included** > checkbox is displayed
8. Click on it 
9. Save
10. Change the shop
11. In the pricing tab > See error: the ecotax is not applied on the second shop

Ping @MatShir, @jolelievre what do you think of those issues, are they blocking this PR?
If no, The PR is QA :heavy_check_mark: 
I tested the full PR, It is  OK :heavy_check_mark: only those two issues.
Automatic tests => OK


Other issues detected but not within the scope of this PR as discussed with @jolelievre, I will describe all issues detected 
**Case1**

1. enable ecotax no tax
2. create a product simple V2
3. Set retail price tax excluded = 10 / no tax rule / retail price tax included = 10 / ecotax = 1
4. create a specific price with a specific price (tax excl.) = 5 from quantity = 1 and save
5. Go to FO > check the product
6. See error the product price = 6 tax included, it should be 5

https://user-images.githubusercontent.com/16067358/170533837-16adfc4c-0f19-4e57-b77b-84eda4ef493c.mp4


**Case2**

1. Go to BO > Taxes > Taxes rule & create a tax rule with combine 10% + 20% = 30%
2. In the BO > Taxes > Enable ecotax with tax rule = combine de tax
3. Go to BO > Create a product V2 > In the pricing tab
4. The ecotax rate displayed = 20% => NOK / it should be 30%
5. also, if I add ecotax = 1 tax exluded => it is displayed 1.2 tax included  / it should be 1.3 tax included

https://user-images.githubusercontent.com/16067358/170534006-1a35ae67-a97a-4be6-a3b2-e8291217cd71.mp4

Same issue for the **combinations** edition modal


**Case3**

1. Add the Arabic language or Algeria pack
2. change the employee language to Arabic
3. Go to BO > Product V2 > Add a new product
4. In the pricing tab, the currency is not displayed: reproduced without PR also (**first** issue)
5. try to change the retail price excluded or included to 10 for example
6. click outside => and the price is converted to .  (second issue)
7. the ( ) are not well displayed

https://user-images.githubusercontent.com/16067358/170534514-fc7beac7-1a9e-4980-aabb-55aecb808cf1.mp4

![image](https://user-images.githubusercontent.com/16067358/170534545-f0302dde-6d23-4619-b3ca-711435bcd6c6.png)

**Case4**

1. change employee language to Arabic
2. edit a product with combinations V2 that have impact on price for each combination
3. In the combinations list => the Final price (tax excl. is NOK
![image](https://user-images.githubusercontent.com/16067358/170534639-23135660-1f4d-43ab-b06c-c2e963cc78e8.png)

**Case5**

1. enable ecotax
2. create a product V2 with retail price tax excluded = 10 / tax rule = 10% /  retail price tax included = 11 and  and save
3. set ecotax = 11
4. enable Display retail price per unit
5. set a Retail price per unit (tax excl.) = 20 per unit / Retail price per unit (tax incl.) = 22 per unit
6. save
7. Go to FO > check the product
8. See error the price per unit displayed is 20 per unit tax included

https://user-images.githubusercontent.com/16067358/170534836-f99ded3c-6c86-4ff0-83f4-a47d5ec13dfe.mp4
![image](https://user-images.githubusercontent.com/16067358/170534870-5937f024-7ac6-4f52-ae76-2677d877e7d9.png)

**Related Case**
1. Create a product V2 with price tax included = 5 / tax rule = 10% / retail price per unit = 10 euro per test / ecotax = 3
![image](https://user-images.githubusercontent.com/16067358/170536289-2389cda3-bf5d-49c6-bdbe-1ff4c833fc34.png)
2. Go to FO => View product => I have this
![image](https://user-images.githubusercontent.com/16067358/170536349-9b88deb0-0739-4ff0-a243-475673a5a01f.png)
I tried the exact prices with **PS1786**
![image](https://user-images.githubusercontent.com/16067358/170536436-b4836ea9-773d-48ce-96f6-cc25225841cf.png)
In the FO => I have this 
![image](https://user-images.githubusercontent.com/16067358/170536484-36d4f567-5042-4457-8f18-a0ac7ba3e158.png)

**Case6**
1. Go to BO > Catalog > Product V2
2. Create a product with combinations > In the combinations list > select all combinations & edit with bulk action
3. Click on Price
4. Set all fields with empty values (just remove the 0) 
![image](https://user-images.githubusercontent.com/16067358/170535508-ab50985a-710d-476a-ac0f-2f59b90c5848.png)
5. Click on apply 
![image](https://user-images.githubusercontent.com/16067358/170535526-c394cf72-af87-4cef-9f75-66a010bec813.png)


**Case7**
1. Go to BO > Catalog > Product V2
2. Create a product with combinations > In the combinations list > select all combinations & edit with bulk action
3. Click on Price
4. Set the impact on price tax excluded
5. See error: the impact on price tax included is not automatically changed 
![image](https://user-images.githubusercontent.com/16067358/170535830-7e02f13b-4a86-486a-9c83-353abe43b176.png)

**Case8**

1. Create pack product V2
2. edit standard page
3. add a product to the pack & save
4. Edit this product with V2
5. Save
6. See exception 
![image](https://user-images.githubusercontent.com/16067358/170535968-5e7e3732-04b9-491e-86a7-56f091fb00bb.png)


If this PR is merged, I will create all issues described above.

Thanks!

jolelievre committed:
Handle ecotax value in grid sorting and filtering when it is enabled

jolelievre committed:
Remove position filter field when column is not displayed

jolelievre committed:
Handle filter by position in product list

jolelievre committed:
Fix product controller tests because filter selector was changed

khouloudbelguith reviewed (commented):
Hi @jolelievre,

I found those issues:

**First issue**

1. Go to BO > Catalog > Product 
2. Add a product V2 > In the options tab: select a supplier and save
3. In the pricing tab > delete the cost price and click outside
4. See error in the console

https://user-images.githubusercontent.com/16067358/173112427-8c58c442-164a-48fa-a023-ffdc7f7debd4.mp4

**Second issue**
1. Go to BO > Catalog > Product 
2. Add a product V2 virtual
3. In the pricing tab > Set the retail price = -1
4. Click on the type product to change the type to a product with combinations
5. See error 

I attached a screen record

https://user-images.githubusercontent.com/16067358/173112935-63d3d882-a0f0-4b4d-a807-d52b3b62b02a.mp4

**Third issue**
1. Go to BO > Catalog > Product 
2. Add a product V2 (simple - combinations - pack)
3. In the pricing tab > set `Retail price (tax excl.) = 1` & `Ecotax (tax excl.) = 2`
4. Change the product type to virtual
5. See error (**first error**)
![image](https://user-images.githubusercontent.com/16067358/173113928-df3a0810-cfd2-4626-a502-a8b96323f284.png)
6. change the `Retail price (tax excl.)` from -1 to 1 and save
7. See error (**second issue**) => the `Retail price (tax excl.)` is converted to 3

I attached a screen record

https://user-images.githubusercontent.com/16067358/173114575-eee4d330-8dae-4a97-ad24-d70051980983.mp4

Thanks!

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationPriceImpactType.php:
```diff
@@ -74,48 +117,125 @@ public function __construct(
+        if ($this->isEcotaxEnabled) {
+            if ($this->isTaxEnabled) {
+                $ecotaxRate = $this->taxComputer->getTaxRate(new TaxRulesGroupId($this->ecoTaxGroupId), new CountryId($this->contextCountryId));
+                $helpMessage = $this->trans('Ecotax rate %rate%%', 'Admin.Catalog.Feature', ['%rate%' => $ecotaxRate->round(2)]);
```
```suggestion
                $helpMessage = $this->trans('Ecotax rate %rate%', 'Admin.Catalog.Feature', ['%rate%' => $ecotaxRate->round(2)]);
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Product/HeaderType.php:
```diff
@@ -92,11 +100,14 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     'data-combinations-warning' => $this->trans('This will delete all combinations.', 'Admin.Catalog.Notification'),
                     'data-pack-warning' => $this->trans('This will delete the list of products in this pack.', 'Admin.Catalog.Notification'),
                     'data-virtual-warning' => $this->trans('This will delete the associated virtual file.', 'Admin.Catalog.Notification'),
+                    'data-ecotax-warning' => $this->trans('This will reset the ecotax value and may impact your retail price tax included.', 'Admin.Catalog.Notification'),
```
```suggestion
                    'data-ecotax-warning' => $this->trans('This will reset the ecotax value and may impact your retail price (tax incl.).', 'Admin.Catalog.Notification'),
```

jolelievre replied:
Thanks @Julievrz it's fixed

MatShir commented:
> Hi @jolelievre,
> 
> I found those issues:
> 
> **First issue**
> 
> 1. Go to BO > Catalog > Product
> 2. Add a product V2 > In the options tab: select a supplier and save
> 3. In the pricing tab > delete the cost price and click outside
> 4. See error in the console
> 
>  Untitled_.Jun.10.2022.5_39.PM.mp4 

To be fixed in this PR 

> **Second issue**
> 
> 1. Go to BO > Catalog > Product
> 2. Add a product V2 virtual
> 3. In the pricing tab > Set the retail price = -1
> 4. Click on the type product to change the type to a product with combinations
> 5. See error
> 
> I attached a screen record
> 
>  Untitled_.Jun.10.2022.5_43.PM.mp4 
> **Third issue**
> 
> 1. Go to BO > Catalog > Product
> 2. Add a product V2 (simple - combinations - pack)
> 3. In the pricing tab > set `Retail price (tax excl.) = 1` & `Ecotax (tax excl.) = 2`
> 4. Change the product type to virtual
> 5. See error (**first error**)
>    ![image](https://user-images.githubusercontent.com/16067358/173113928-df3a0810-cfd2-4626-a502-a8b96323f284.png)
> 6. change the `Retail price (tax excl.)` from -1 to 1 and save
> 7. See error (**second issue**) => the `Retail price (tax excl.)` is converted to 3
> 
> I attached a screen record
> 
>  Untitled_.Jun.10.2022.5_54.PM.mp4 
> Thanks!

For the bugs around the type changement, I suggest making a dedicated EPIC, so we can fix them at once. It is not related to this PR anyway.

jolelievre committed:
Refacto FormObjectMapper no longer relies on eventEmitter any more, only use watchers now and internal events that can be stopped

jolelievre committed:
Remove useless mapper events, and stop propagation when 0 value is forced after invalid input

jolelievre commented:
Hi @khouloudbelguith 

the first issue has been fixed in this PR. For the other ones here is the Epic created by @MatShir https://github.com/PrestaShop/PrestaShop/issues/28738

jolelievre committed:
Fix wording

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Pricing/RetailPriceType.php:
```diff
@@ -127,28 +152,50 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 ],
                 'label' => $this->trans('Tax rule', 'Admin.Catalog.Feature'),
-                'help' => !$this->taxEnabled ? $this->trans('Tax feature is disabled, it will not affect price tax included.', 'Admin.Catalog.Feature') : '',
+                'help' => !$this->isTaxEnabled ? $this->trans('Tax feature is disabled, it will not affect price tax included.', 'Admin.Catalog.Feature') : '',
```
It's weird. If the tax is disabled, it seems clear that the price (including tax) will not be affected... but neither will the price excluding tax. If there is no tax? I find this message confusing and don't understand what the message is.

jolelievre replied:
I don't know it's been here for a while, it was not the purpose of the PR, only the variable name was changed If you think it's useless we can clear it But maybe in another I don't want to block this one because of unrelated wording

If I have other bug to fix in the PR I will do it, if not we can clear it when we'll perform the wording review on the whole page

khouloudbelguith reviewed (approved):
Hi @jolelievre,

Automatic tests are ok :heavy_check_mark: 
It is ok :heavy_check_mark: 

When we merge the PR, we need to create all old issues detected.

Thanks!

jolelievre commented:
Thanks @khouloudbelguith
